File: 4th Semester (Java)/lab-1/src/main/java/transactions/WithdrawTransaction.java
package transactions;

import accounts.Account;
import exceptions.BanksException;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Represents a transaction of withdrawing money from account.
 */
public class WithdrawTransaction extends Transaction {
    /**
     * WithdrawTransaction constructor
     * @param account the account transaction was made from
     * @param moneyAmount amount of money in transaction
     * @param commission commission amount was taken from client
     * @param dateTime date and time of transaction
     */
    public WithdrawTransaction(Account account, double moneyAmount, double commission, LocalDateTime dateTime) {
        Id = UUID.randomUUID();
        this.account = account;
        this.moneyAmount = moneyAmount;
        this.commission = commission;
        this.dateTime = dateTime;
    }

    /**
     * Cancels transaction and returns money to account.
     * @throws BanksException if transaction is already canceled
     */
    public void cancel() throws BanksException {
        if (canceled) {
            throw new BanksException("Failed to Cancel transaction, transaction is already canceled!");
        }

        double newMoneyAmount = account.getMoney() + moneyAmount + commission;
        account.setMoney(newMoneyAmount);
        canceled = true;
    }
}

File: 3rd Semester (C#)/Lab4/Banks/Models/WithdrawTransaction.cs
using Banks.Interfaces;
using Banks.Tools;

namespace Banks.Models;

public class WithdrawTransaction : ITransaction
{
    public WithdrawTransaction(IAccount account, double moneyAmount, double comission, DateTime dateTime)
    {
        Id = Guid.NewGuid();
        Account = account;
        MoneyAmount = moneyAmount;
        Comission = comission;
        DateTime = dateTime;
    }

    public Guid Id { get; }
    public IAccount Account { get; }
    public double MoneyAmount { get; }
    public double Comission { get; }
    public DateTime DateTime { get; }
    public bool Canceled { get; private set; }

    public void Cancel()
    {
        if (Canceled)
        {
            throw new BanksException($"Failed to Cancel transaction: {this}, transaction is already canceled!");
        }

        Account.Money += MoneyAmount + Comission;
        Canceled = true;
    }
}
